submariner/cable-driver-policy.md

```diff
@@ -0,0 +1,165 @@
+# Cable Driver Policy Support
+
+## Summary
+
+Submariner currently offers three cable drivers: VXLAN, IPSec or wireguard. However only one of these cable drivers
+is enabled across all the clusters in a ClusterSet; That is all clusters that join a broker must use the same cable
+driver. The goal of this proposal is to extend the current model to support the selection of different cable types
+between connected clusters.
+
+There are use cases that involve supporting a mix of cable drivers or cable driver configurations within the same
+ClusterSet. For example, a ClusterSet could contain three clusters, two that are on-premise and the other
+in a public cloud. Connections to the cloud cluster will need to use encryption while the on-premise clusters can use
+unencrypted connections, as they’re connected by a private link, to avoid the performance overhead of encryption.
+
+## Proposal
+
+This enhancement proposes a new policy to allow users to selectively deploy different cable types between clusters
+in a ClusterSet. The proposal assumes that connection between two clusters is symmetrical. If Cluster A is connected
+to Cluster B via a VXLAN cable type - then Cluster B is connected to Cluster A via a VXLAN cable type and not another
+cable type.
+
+This enhancement will propose a single API for cable driver policy support using custom resource definition (CRDs).
+
+### Cable driver policy API
+
+This following CRD is proposed:
+
+```Go
+type ClusterConnectionPolicies struct {
+    metav1.TypeMeta `json:",inline"`
+
+    metav1.ObjectMeta `json:"metadata,omitempty"`
+
+    ClusterConnectionPoliciesSpec `json:"spec"`
+}
+
+type ClusterConnectionPoliciesSpec struct {
+ 
+    Policies []ClusterConnectionPolicy `json:"policies,omitempty"`
```

Comment:
```Go
type ClusterConnectionPolicy struct {
    metav1.TypeMeta `json:",inline"`

    metav1.ObjectMeta `json:"metadata,omitempty"`

    ClusterConnectionPolicySpec `json:"spec"`

    // Status is the current state of the ClusterConnectionPolicy.
    Status ClusterConnectionPolicyStatus `json:"status,omitempty"`
}

type ClusterConnectionPolicySpec struct {
    // LeftClusterSelector identifies the Cluster resources representing the clusters on one end of a connection. An empty
    // selector indicates wildcard, ie matches any cluster.
    LeftClusterSelector metav1.LabelSelector `json:"leftClusterSelector,omitempty"`

    // RightClusterSelector identifies the Cluster resources representing the clusters on the other end of a connection.
    // An empty selector indicates wildcard, ie matches any cluster.
    RightClusterSelector metav1.LabelSelector `json:"rightClusterSelector,omitempty"`

    // Name of the cable driver implementation to use for this connection.
    CableDriverName string `json:"CableDriverName"`
}

// ClusterConnectionPolicyConditionType is the type for status conditions on
// a ClusterConnectionPolicy This type should be used with the
// ClusterConnectionPolicyStatus.Conditions field.
type ClusterConnectionPolicyConditionType string

const (
  // ClusterConnectionPolicyConditionStatusApproved represents status of a ClusterConnectionPolicy that is approved
  // the ClusterConnectionPolicy provider and will be implemented in the cluster
  ClusterConnectionPolicyConditionStatusApproved ClusterConnectionPolicyConditionType = "Approved"

  // ClusterConnectionPolicyConditionStatusFailure represents status of a ClusterConnectionPolicy that is not approved
  // ClusterConnectionPolicy provider and will not be implemented in the cluster
  ClusterConnectionPolicyConditionStatusFailure ClusterConnectionPolicyConditionType = "Failure"
)

// ClusterConnectionPolicyConditionReason defines the set of reasons that explain why a
// particular ClusterConnectionPolicy condition type has been raised.
type ClusterConnectionPolicyConditionReason string

const (
  // ClusterConnectionPolicyConditionReasonPolicyError represents a reason where the ClusterConnectionPolicy may not have been
  // implemented in the cluster due to an issue detected by the ClusterConnectionPolicy provider with another policy
  ClusterConnectionPolicyConditionReasonPolicyError ClusterConnectionPolicyConditionReason = "PolicyError"
)

// ClusterConnectionPolicyStatus describe the current state of the ClusterConnectionPolicy.
type ClusterConnectionPolicyStatus struct {
  // Conditions holds an array of metav1.Condition that describe the state of the ClusterConnectionPolicy.
  Conditions []metav1.Condition `json:"conditions,omitempty"`
}
```

Reply:
yeah we could take the fail fast approach to conflicts/ambiguity. I'm not sure tho we should automatically fallback to the default policy. Connectivity may still **seem** to be working to the user but they might not be aware that their intended policy didn't get applied. 


+1 on adding status. However it wouldn't be a singular status but one per remote cluster, ie the policy may be successfully applied for one cluster but not another.  Also the `ClusterConnectionPolicy` CRD is synced from the broker so we'd have to think about those dynamics, ie do we update the local synced copy or create another local CRD to store the status.

Reply:
The intention for the status in `ClusterConnectionPolicy` to reflect the status of the connection policy rather than its application. So it's either approved or declined (I will change the terminology from failed)... We could do something along the lines of only approved policies will be synced to clusters from the broker and a GW will only apply approved policies...

The `ConnectionStatus` is already available per cluster pair through the GW status which looks like it's also already synced I think(if we do a subctl show all for example it shows the status of GW connectivity)? I'm not sure if we want to duplicate that.

```go
type GatewayStatus struct {
	Version       string       `json:"version"`
	HAStatus      HAStatus     `json:"haStatus"`
	LocalEndpoint EndpointSpec `json:"localEndpoint"`
	StatusFailure string       `json:"statusFailure"`
	Connections   []Connection `json:"connections"`
}

type Connection struct {
	Status        ConnectionStatus `json:"status"`
	StatusMessage string           `json:"statusMessage"`
	Endpoint      EndpointSpec     `json:"endpoint"`
	UsingIP       string           `json:"usingIP,omitempty"`
	UsingNAT      bool             `json:"usingNAT,omitempty"`
	// +optional
	LatencyRTT *LatencyRTTSpec `json:"latencyRTT,omitempty"`
}

type ConnectionStatus string

const (
	Connected       ConnectionStatus = "connected"
	Connecting      ConnectionStatus = "connecting"
	ConnectionError ConnectionStatus = "error"
)
```

The only concern I have about not connecting/fallingback to the default driver is that this is not what submariner has done historically. The clusters are always connected in a full mesh fashion. If we only sync approved policies to all `clusters` in a `ClusterSet` - when a policy is declined a cluster will never see it and fall back to the default connection policy anyway. When the updated policy is approved - the connection would need to be restarted between a pair of clusters...

But as per your suggestion, then we would need to sync all policies regardless of their status (approved or declined) then we can extend the ConnectionStatus to include waiting for policy? or potentially just fallback to error and then there's no need for a new ConnectionStatus. But this introduces the concept of no connection existing between a pair of clusters.

```go
type ConnectionStatus string

const (
	Connected       ConnectionStatus = "connected"
	Connecting      ConnectionStatus = "connecting"
	ConnectionError ConnectionStatus = "error"
	ConnectionWaiting ConnectionStatus = "waiting"
)
```

Reply:
I thought conflict could be detected much earlier on - via a controller or the syncer on the broker, covering both the pre-check and post deployment policy creation (as I also thought we are going to limit policy creation to the broker only and sync one way rather than bi-directionally)... And then I even thought that maybe a validation webhook might even be the way to go (build in the logic to detect conflicts much earlier on, this way only conflict free policies are approved for admission) so issues will be reported pretty quickly to the user...

I think if we are getting to the point of trying to establish a connection and then detecting a policy conflict we're a little late to the party. There could be many conflicting policies already causing no connections across the `ClusterSet`. Say for example (bear with me) an insider bad actor creates a bunch on conflicting policies to result in a set of no connections across your `ClusterSet`. Or even an unintended consequence could be for users to deliberately create conflicting policies to allow them to have no connections between clusters in a ClusterSet.

I think that conflicting policies is something we would like to tackle as early as possible. Maybe the validation webhook is the way to go rather than relying on additional controllers or syncers to detect conflicts.

Also - maybe we want to encourage the users to use ipsec driver in the default policy - as the most `secure` driver.  Then build the VXLAN connections on top of that. That way they will always fall back to an encrypted connection between clusters.

Reply:
> I thought conflict could be detected much earlier on 

That would be ideal but, as I mentioned, conflicts could occur due to a mislabeled cluster, not just on policy creation/update. Let's consider an example.

I create 2 policies:

```yaml
leftClusterSelector:
  matchLabels:
    org: sales
rightClusterSelector:
  matchLabels:
    org: sales
cableDriverName: "driver1"

leftClusterSelector:
  matchLabels:
    org: engineering
rightClusterSelector:
  matchLabels:
    org: engineering
cableDriverName: "driver2"
```

The default cable driver is "_ipsec_".

There's no clusters labeled yet so no conflicts - the policies are approved, synced and all is good. Some time later, clusters _sales1_ and _sales2_  are labeled with `org: sales`. Only the first policy matches so there's no conflicts  and the GWs reconnect using "_driver1_". Clusters _engineering1_ and _engineering2_  are labeled with `org: engineering` and are reconnected using "_driver2_". Then _sales-and-eng1_ is labeled with `org: sales` and `org: engineering`. When connecting to  _sales1_ and _sales2_  only the first policy matches and when connecting to  _engineering1_ and _engineering2_ only the second policy matches. Again no conflicts. Then _sales-and-eng2_ is also labeled with `org: sales` and `org: engineering`. Now, when attempting to connect with _sales-and-eng1_, both policies match and there's a conflict.

So what remediation steps should be taken... Should we un-approve/revoke a policy? If so, which one? Or both? But what about the connectivity between _sales1_ and _sales2_? There's no conflict there so if the first policy is  revoked then they would implicitly and silently fall back to using "_ipsec_".  Similar with _engineering1_ and _engineering2_. Is it even the policies that need to be corrected or is it a mislabeled cluster?

I think this example illustrates how doing upfront checking and and rejecting policies can be problematic, especially since conflicts are per connection. 

We could do validation of a policy wrt the label selection syntax and the cable driver name.

> via a controller or the syncer on the broker, covering both the pre-check and post deployment policy creation  
> (as I also thought we are going to limit policy creation to the broker only and sync one way rather than bi-directionally)... 

yeah we sync one way, ie member cluster to or from the broker cluster. We've strived to keep the broker simple, ie no controllers although we did add an operator controller running on the broker. But that doesn't preclude us from adding another controller if there's a compelling need. 

> And then I even thought that maybe a validation webhook might even be the way to go 

yeah that would seem preferable over a controller for whatever validation we want to do.

> I'm not sure what the best approach is here, early detection would certainly avoid unintended consequences ... 

As outlined above, early detection would be preferable but it doesn't seem feasible in all cases. I'd think the common scenario would be that policies are created up front with conflicts occurring when clusters are subsequently labeled.

> Later detection could be simpler but if we don't fall back to the default connection policy then could result in some unintended consequences/uses of the no connect feature.

yeah perhaps you're right. We still need a way to report/record conflicts, perhaps update the `StatusMessage` in the connection status at the very least.

Reply:
I see your point on the early checking not covering all scenarios.

And even with the no-connection and marking the status of a connection as a conflict - is certainly a better approach then a full fall back to a default driver and no error reporting...

I will write up an update based on the discussion points above. and then we can review and see what else is needed :)